tabgroups/init.lua

```diff
@@ -0,0 +1,870 @@
+    { "<Return>", "Select tabgroup.", select_tabgroup }
+}, menu_binds))
+
+new_mode("tabgroup-menu", {
+    enter = function (w, tabgroup_name)
+        local rows = build_tabgroup_menu(w, tabgroup_name)
+        _operation = nil
+        w.menu:build(rows)
+        w:set_input()
+        local notify = _default_notify
+        if tabgroup_name then
+            if number_of_tabgroups(w) > 1 then
+                notify = notify ..", ".."m: move selected tab to another tabgroup, -: show list of groups"
+            else
+                notify = notify ..", ".."-: show list of groups"
+            end
+        else
+            notify = notify ..", ".. "+: show tabs in selected group"
+        end
+        w:notify(notify, false)
+    end,
+
+    leave = function (w)
+        w.menu:hide()
+    end,
+})
+
+add_binds("tabgroup-menu", lousy.util.table.join({
+    { "<Return>", "Switch to tabgroup or tab in tabgroup.", switch_tabgroup_or_tab },
+    { "n", "Create new tabgroup.", new_tabgroup },
+    { "r", "Rename tabgroup.", rename_tabgroup },
+    { "d", "Delete tabgroup.", function(w)
+        local row = w.menu:get()
+        w:set_mode()
+        if row and row._group then
+            local deleted = delete_tabgroup(w, row._group)
+            open_tabgroup_menu(w)
+            if not deleted then
+                w:notify("Can't remove last tabgroup")
+            end
+        end
+    end},
+    { "+", "Open list of tabs in tabgroup.", show_tabgroup_content },
+    { "-", "Hide list of tabs in tabgroup.", show_tabgroups },
+    { "m", "Move tab to different tabgroup.", move_tab_to_tabgroup_menu },
+    { "1", "Switch to first tab or tabgroup.", switch_tabgroup_or_tab, { count = 1 } },
+    { "2", "Switch to second tab or tabgroup.", switch_tabgroup_or_tab, { count = 2 }  },
+    { "3", "Switch to third tab or tabgroup.", switch_tabgroup_or_tab, { count = 3 }  },
+    { "4", "Switch to forth tab or tabgroup.", switch_tabgroup_or_tab, { count = 4 }  },
+    { "5", "Switch to fifth tab or tabgroup.", switch_tabgroup_or_tab, { count = 5 }  },
+    { "6", "Switch to sixth tab or tabgroup.", switch_tabgroup_or_tab, { count = 6 }  },
+    { "7", "Switch to seventh tab or tabgroup.", switch_tabgroup_or_tab, { count = 7 }  },
+    { "8", "Switch to eighth tab or tabgroup.", switch_tabgroup_or_tab, { count = 8 }  },
+    { "9", "Switch to ninth tab or tabgroup.", switch_tabgroup_or_tab, { count = 9 }  },
+}, menu_binds))
+
+add_binds("normal", {
+    { "x", "Open tabgroup menu.", open_tabgroup_menu },
+    { "X", "Show tabs in active tabgroup.", function (w) open_tabgroup_menu(w, w2groups[w].active)  end },
+})
+
+settings.register_settings({
+    ["tabgroups.sort_groups_by"] = {
+        type = "string",
+        default = "name asc",
+        validator = function(v)
+            local field, order = string.match(v, "^%s*(%w+)%s+(%w+)")
+            if field then
+                field = string.lower(field)
+                if field ~= 'name' and
```

Comment:
- `field` and `order` are either both set or both nil
- if they're both nil, that's an error condition: return false
```lua
if not field then return false end
field, order = field:lower(), order:lower()
field = ({name = true, atime = true, mtime = true, ctime = true, hits = true})[field]
order = order == "asc" or order == "desc"
return field and order
```